kernel/arch/dreamcast/hardware/biosfont.c

```diff
@@ -354,16 +370,39 @@ void bfont_draw_str_ex(void *b, uint32 width, uint32 fg, uint32 bg, uint8 bpp, u
     }
 }
 
-void bfont_draw_str(void *b, uint32 width, uint8 opaque, const char *str) {
-    bfont_draw_str_ex(b, width, bfont_fgcolor, bfont_bgcolor, (bfont_32bit ? (sizeof (uint32)) : (sizeof (uint16))) << 3, opaque, str);
+/* Draw string of full-width (wide) and half-width (thin) characters
+   Note that this handles the case of mixed encodings unless Japanese
+   support is disabled (BFONT_CODE_ISO8859_1).
+   XXX: Seems like this can be shrunk to use uint8_t for nChr/Mask/Flag and
+    getting rid of nMask.
+   */
+void bfont_draw_str_ex(void *b, uint32_t width, uint32_t fg, uint32_t bg, uint8_t bpp, 
+                       bool opaque, const char *fmt, ...) {
+    va_list var_args;
+    va_start(var_args, fmt);
+
+    bfont_draw_str_ex_va(b, width, fg, bg, bpp, opaque, fmt, &var_args);
+
+    va_end(var_args);
+}
```

Comment:
```suggestion
    DONT COMMIT ME OR QUZAR WONT BE ABLE TO REVIEW
int bfont_draw_printf(void *b, uint32_t width, uint32_t fg, uint32_t bg, uint8_t bpp, 
                       bool opaque, const char *fmt, ...) {
    char string[1088];
    int rv = 0;
    va_list var_args;
    va_start(var_args, fmt);
    rv = vsnprintf(string, sizeof(string), fmt, *var_args);
    va_end(var_args);    
    
    bfont_draw_str_ex(b, width, fg, bg, bpp, opaque);
    return rv;
}
```

Something like this is what I'm suggesting. It's almost identical to how dbgio does it, and this way you can even provide the return from vsnprintf so that the user can know if they're going to be truncated or something.

Reply:
K, done! I'm not returning the return value from vsnprintf(), though, because it's how big of a buffer you NEED to have to not be truncated, and it's not like they have any control over that buffer either, because it's internal. Seemed pointless.